File: 옛날에 푼거 정리/bak9012.cpp
//()괄호 맞추기

#include <stdio.h>
#include <string.h>

int main() {

	int i, j, n, size, num;
	char s[50];

	scanf("%d\n", &n);

	for (i = 0; i < n; i++) {
		num = 0;
		scanf("%s", s);
		
		size = strlen(s);
;
		for (j = 0; j < size; j++) {
			if (s[j] == '(') {
				num ++;
			}
			else if (s[j] == ')' && num > 0) {	//else if 케이스 두개 나눠주는거 ! (한번 틀림)
				num--;
			}
			else if (s[j] == ')' && num <= 0) {
				num--;
				break;
			}
		}

		if (num == 0) {
			printf("YES\n");
		}
		else {
			printf("NO\n");
		}

	}

}

File: python/bak-step17/bak9012.py
#9012
#괄호
import sys
for i in range(int(sys.stdin.readline())):
     TF = True
     num = 0
     tmp = sys.stdin.readline()
     for j in range(len(tmp)):
          if num<0:
               TF=False
               break
          if tmp[j] == '(':
               num+=1
          else:
               num-=1
     if num == -1:
          if TF:
               print("YES")
          else:
               print("NO")
     else:
          print("NO")
